opamp/observiq/agent_config_manager.go

```diff
@@ -0,0 +1,215 @@
+		configContents, err := os.ReadFile(cleanPath)
+		if err != nil {
+			return nil, fmt.Errorf("error reading config file %s: %w", configName, err)
+		}
+
+		// Add to contentMap
+		contentMap[configName] = &protobufs.AgentConfigFile{
+			Body:        configContents,
+			ContentType: opamp.DetermineContentType(managedConfig.ConfigPath),
+		}
+	}
+
+	return &protobufs.EffectiveConfig{
+		ConfigMap: &protobufs.AgentConfigMap{
+			ConfigMap: contentMap,
+		},
+	}, nil
+}
+
+// ApplyConfigChanges compares the remoteConfig to the existing and applies changes
+func (a *AgentConfigManager) ApplyConfigChanges(remoteConfig *protobufs.AgentRemoteConfig) (effectiveConfig *protobufs.EffectiveConfig, changed bool, returnErr error) {
+	// Always compute effective config at the end. This ensures we always have the most up to date copy when we respond to the server
+	defer func() {
+		var err error
+		effectiveConfig, err = a.ComposeEffectiveConfig()
+		if err != nil {
+			a.logger.Error("Failed to compute effective config while applying config changes", zap.Error(returnErr))
+		}
+	}()
+
+	remoteConfigMap := remoteConfig.GetConfig().GetConfigMap()
+
+	// No remote config Map
+	if remoteConfigMap == nil {
+		return
+	}
+
+	// loop through all remote configs and compare then with existing configs
+	for configName, remoteContents := range remoteConfigMap {
+		// For security check the log file we want is acceptable
+		if _, ok := acceptableConfigs[configName]; !ok {
+			a.logger.Warn("Not supported config received skipping", zap.String("config", configName))
+			continue
+		}
+
+		managedConfig, ok := a.configMap[configName]
+		// This check is impossible to hit now but will be a use case we'll have in the future so leaving this tested code in for now.
+		if !ok {
+			// We don't current track this config file we should add it
+			if err := a.trackNewConfig(configName, remoteContents.GetBody()); err != nil {
+				returnErr = err
+				return
+			}
+			changed = true
+			continue
+		}
+
+		// Update the config file
+		configChanged, err := a.updateExistingConfig(configName, managedConfig, remoteContents.GetBody())
+		if err != nil {
+			returnErr = err
+			return
+		}
+
+		changed = changed || configChanged
+	}
+
+	return
+}
+
+func (a *AgentConfigManager) updateExistingConfig(configName string, managedConfig *opamp.ManagedConfig, newContents []byte) (changed bool, err error) {
+	remoteHash := opamp.ComputeHash(newContents)
+
+	// Nothing to update
+	if bytes.Equal(managedConfig.GetCurrentConfigHash(), remoteHash) {
+		// Verify disk matches memory and remote. This will ensure if any changes were made to the config on disk
+		// the will be overwritten by what is expected.
+		return verifyDiskContents(managedConfig.ConfigPath, managedConfig.GetCurrentConfigHash(), newContents)
+	}
+
+	a.logger.Info("Applying changes to config file", zap.String("config", configName))
+	changed, err = managedConfig.Reload(newContents)
```

Comment:
Yeah I kept the word changed since that's what it is in the opamp library. They call it `configChanged` so I wanted to keep the terminology.

>it doesn't actually represent that anything was changed. just that the file has properly been overwritten with the remote config data

This is not always true and it depends on the reload function. All of our reload functions do a hot reload so in the case of the manager it changes the identity of the agent and for the logger it changes the logger object. 

I think adding a comment though to the main struct is not a bad idea just to make it clear.